### PR TITLE
Allow the default block pool to be overridden

### PR DESCRIPTION
**Pull request by joshlory**
Creating a new `level_type` in the Block Pool:

screen shot 2018-07-18 at 9 57 52 am

Loading a non-default Block Pool in a given level:

image

### CODE REVIEW

**Comment by balderdash** [comment]
I don't think we should overload the `level_type` field to refer to both the level type and identify the block pool. I expect every block will be built for a specific level type, so we should still store that data.

I can think of two options:
* Just add a `block_pool` field to the block, and you can specify that in the URL to load a different set of blocks.
* Create a `blook_pools` (or perhaps `level_subtypes`) table to store the `level_type` associated with a block pool, and replace the block's `level_type` with a `block_pool`/`level_subtype` field.
  * This would also be the place to specify a different helper library
  * We could even add a `parent_block_pool` field to this table and support arbitrary type hierarchies. With a little work that would allow levelbuilders to fork their own version of spritelab without help from an Engineer.

The first option is much easier, and could pretty easily be converted to the second if you don't want to put in the work for that just yet.

**Comment by mrjoshida** [comment]
+1 to the heirarhy idea - I think in many (most?) cases we'll want to start with an existing base block pool to add on top of, even if we eventually replace that pool with customized alternatives.

**Comment by joshlory** [comment]
I'm planning to allow multiple sandboxes to be specified, for example:

image

Does that cover your use case @mrjoshida?

My thought process is that Block Pools should be cheap to create.  What defines an "App" is a Visualization (e.g. the Maze renderer or Sprite Lab game renderer) + 1 or more block sets + (maybe) a helper API.

**Comment by joshlory** [comment]
In this world, a `level_type` would just mean which levels gets a given Block Pool by default.  You can always mix Block Pools into other levels as needed.

**Comment by joshlory** [comment]
My bias is it seems like we've been shifting away from re-using Levels, etc. to creating a unique copy per curriculum+year.  I'm making the assumption that prototyping new blocks is common, but when we cut a new release of the toolbox _ALL_ blocks get versioned.

**Comment by mrjoshida** [comment]
That sounds great

**Comment by balderdash** [comment]
That seems reasonable, allowing arbitrary sets of block pools makes it easy to transfer blocks between worlds, and doesn't preclude an app type hierarchy.

I still think `level_type` and `block_pool` should be separate, with the default block pool indicated by an empty/special block pool name. I can't think of any reason you would want to transfer blocks between level types, but finding all the blocks for a level type could be useful.

**Comment by joshlory** [comment]
> I can't think of any reason you would want to transfer blocks between level types

I'm thinking this will actually become _more_ common, as we extract useful library blocks to a shared pool.

**Comment by balderdash** [comment]
We'll transfer blocks between "subtypes", but not between levels based on gamelab jr and levels based on maze.

**Comment by joshlory** [comment]
There were enough changes that would've caused conflicts, I went ahead and finished up the editing and loading blocks with multiple pools.  Lmk if you want me to extract out any parts of this PR to reduce the diff!

edit-block-pool

**Comment by joshlory** [comment]
[Different sub-types / sandboxes for Sprite Lab, where blocks are in separate universes](https://trello.com/c/fWwgtcfm/221-different-sub-types-sandboxes-for-sprite-lab-where-blocks-are-in-separate-universes)

**Comment by joshlory** [comment]
@balderdash PTAL at the changes since you last looked!

File: dashboard/app/views/blocks/edit.html.haml [hunk 1]
```diff
@@ -12,8 +12,8 @@
       url: @block.new_record? ? blocks_path : block_path(id: @block.name),
       html: {id: 'edit-block-form', method: @block.new_record? ? 'post' : 'patch'},
     ) do |f|
-      = f.label 'Level Type'
-      = f.select :level_type, ['GamelabJr', 'Maze']
+      = f.label 'Pool Name'
+      = f.select :pool, Block.all_pool_names + [params[:pool]].compact
```
Comment (balderdash) [inline dashboard/app/views/blocks/edit.html.haml:16]: Is there a GUI for adding a new pool, or do you just have to add this query param?

Comment (joshlory) [reply dashboard/app/views/blocks/edit.html.haml:16]: Only the query param, for now.  I think we'll want a separate model+controller to manage these pools (and what libraries / shared functions / app type they're associated with).

**Pull request merged**